File: C++/Hard/WrongAnswer/concatenated-words_0.cpp
class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        
        unordered_set<string> dict(words.begin(), words.end());
        vector<string> res;
        for (string& word : words) {
            int n = word.size();
            vector<int> dp(n + 1);
            dp[0] = 1;
            for (int i = 1; i <= n; ++i) {
                for (int j = 0; j < i; ++j) {
                    if (dp[j] && dict.count(word.substr(j, i - j))) {
                        dp[i] = 1;
                        break;
                    }
                }
            }
            if (dp[n]) res.push_back(word);
        }
        return res;
    }
};

File: Python3/Hard/TimeLimitExceeded/concatenated-words_0.py
class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        
        words = set(words)
        def dfs(word):
            for i in range(1, len(word)):
                if word[:i] in words and (word[i:] in words or dfs(word[i:])):
                    return True
            return False
        return [word for word in words if dfs(word)]